parsers/ENTSOE.py

```diff
@@ -736,44 +744,72 @@ def parse_scalar(
+    storage = StorageMix()
+    fuel_em_type = ENTSOE_PARAMETER_BY_GROUP[fuel_code]
+    if fuel_code in ENTSOE_STORAGE_PARAMETERS:
+        # Only include consumption if it's for storage. In other cases
+        # it is power plant self-consumption which should be ignored.
+        storage.set_value(fuel_em_type, -quantity)
+        return None, storage
+    if 0 > quantity > -50:
+        logger.info(
+            "Self consumption value %s for %s has been set to 0."
+            % (quantity, fuel_em_type),
+            extra={"key": zoneKey, "fuel_type": fuel_em_type},
+        )
+        quantity = 0
+    production.set_value(fuel_em_type, quantity)
+    return production, None
+
+
 def parse_production(
-    xml_text,
-) -> Union[Tuple[List[Dict[str, Any]], List[datetime]], None]:
-    if not xml_text:
-        return None
-    soup = BeautifulSoup(xml_text, "html.parser")
-    # Get all points
-    productions = []
-    datetimes = []
+    xml: str,
+    logger: Logger,
+    zoneKey: ZoneKey,
+    forecasted: bool = False,
+) -> ProductionBreakdownList:
+    all_production_breakdowns = []
+    source_type = EventSourceType.forecasted if forecasted else EventSourceType.measured
+    if not xml:
+        return ProductionBreakdownList.merge_production_breakdowns(
+            all_production_breakdowns, logger
+        )
+    soup = BeautifulSoup(xml, "html.parser")
+
+    # Each timeserie is dedicated to a different fuel type.
     for timeseries in soup.find_all("timeseries"):
+        production_breakdowns = ProductionBreakdownList(logger)
         resolution = str(timeseries.find_all("resolution")[0].contents[0])
         datetime_start: arrow.Arrow = arrow.get(
             timeseries.find_all("start")[0].contents[0]
         )
-        is_production = (
-            len(timeseries.find_all("inBiddingZone_Domain.mRID".lower())) > 0
-        )
-        psr_type = str(
+        fuel_code = str(
             timeseries.find_all("mktpsrtype")[0].find_all("psrtype")[0].contents[0]
         )
 
         for entry in timeseries.find_all("point"):
             quantity = float(entry.find_all("quantity")[0].contents[0])
             position = int(entry.find_all("position")[0].contents[0])
             datetime = datetime_from_position(datetime_start, position, resolution)
-            try:
-                i = datetimes.index(datetime)
-                if is_production:
-                    productions[i][psr_type] += quantity
-                elif psr_type in ENTSOE_STORAGE_PARAMETERS:
-                    # Only include consumption if it's for storage. In other cases
-                    # it is power plant self-consumption which should be ignored.
-                    productions[i][psr_type] -= quantity
-            except ValueError:  # Not in list
-                datetimes.append(datetime)
-                productions.append(defaultdict(lambda: 0))
-                productions[-1][psr_type] = quantity if is_production else -1 * quantity
```

Comment:
I do like the idea of moving up the validation and such here so perhaps the add_value() method should be implemented here if at all and return a production and storage mix.

Reply:
I'm actually not super sure about it. Since we are looping over all fuel timeseries, I would feel more safe using the merge production outputs to make sure that timestamps are aligned.
In the EIA, we saw that the add method wouldn't work because the timestamps are not all aligned.

Reply:
Don't ENTSO-E TP normalize the times to the MTU resolution in UTC?
So they should all always be for the same hour or same 15 minutes depending on the country. But those should never be mixed.

Reply:
I hope they do it, but to be on the safe side, I rather avoid trusting the source to do transformations for us.
My reasoning here is that we iterate per fuel report, so it seems more efficient to match the data form when parsing it and merging it as a final step. Otherwise it would require adding some logic in the ENTSOE parser to find back the correct date to which we should add this mode. Since we have the logic for merging lists, I would rather use this.

Reply:
Okay that makes sense. I'll take a deep look at the changes in here today.